fix(template_processing): get_filters now works for IS_NULL and IS_NOT_NULL operators

Prokos opened the pull request:
### SUMMARY
The macro `get_filters` within the Jinja context can be used to implement filtering in custom locations in your virtual dataset SQL. This works great, except for when filtering with the IS_NULL or IS_NOT_NULL operator. These are ignored, cannot be used, and will break certain implementations that rely on remove_filter to be functional (like the one I've outlined below).

The reason the code breaks for these operators is because they are unique in the sense that they don't have a comparator. Adjusting the code to not need a comparator for these operator types fixes the issue.


### TESTING INSTRUCTIONS
1. Create a virtual dataset on BigQuery with the following SQL 
- likely reproducible with less - but it's vital that both column selection/grouping and filtering is handled
- not a BigQuery specific bug, thats just the database I use
```
{% set finalColumns = columns if columns is defined else ['__ALL__'] %}

SELECT


  sum(measure) as measure,
  
  {% for col in finalColumns %}

  	{% if col == 'col1' or col == '__ALL__' or (col is mapping and 'sqlExpression' in col and col.sqlExpression == 'col1') %}
  		col1
  		{%- if col == '__ALL__' -%},{%- endif -%}
  		{%- if not loop.last -%},{%- endif -%}
  	{% endif %}
  	
  	{% if col == 'col2' or col == '__ALL__' or (col is mapping and 'sqlExpression' in col and col.sqlExpression == 'col2') %}
  		col2
  		{%- if col == '__ALL__' -%},{%- endif -%}
  		{%- if not loop.last -%},{%- endif -%}
  	{% endif %}
  	
  {% endfor %}
  
FROM UNNEST([
  STRUCT(1 as measure, 'value1A' as col1, 'value2A' as col2),
  STRUCT(2 as measure, 'value1A' as col1, 'value2B' as col2),
  STRUCT(5 as measure, 'value1B' as col1, 'value2B' as col2),
  STRUCT(15 as measure, NULL as col1, NULL as col2)
])

-- FILTERS

{% for filter in get_filters('col1', remove_filter=True) -%}
    {%- if loop.first %} WHERE 1 = 1 {% endif %}
    AND
    {% if filter.get('op') == 'IN' -%}
        col1 IN {{ filter.get('val')|where_in }}
    {% elif filter.get('op') == 'NOT IN' -%}
        col1 NOT IN {{ filter.get('val')|where_in }}
    {% elif filter.get('op') == 'ILIKE' -%}
        LOWER(col1) LIKE LOWER({{ filter.get('val') }})
    {% elif filter.get('op') == 'IS NULL' -%}
        col1 IS NULL
    {% elif filter.get('op') == 'IS NOT NULL' -%}
        col1 IS NOT NULL
    {% else -%}
        col1 {{ filter.get('op') }} {{ filter.get('val') }}
    {%- endif -%}
{%- endfor %}

{% for filter in get_filters('col2', remove_filter=True) -%}
    {% if loop.first %} WHERE 1 = 1 {% endif %}
    AND
    {% if filter.get('op') == 'IN' -%}
        col2 IN {{ filter.get('val')|where_in }}
    {% elif filter.get('op') == 'NOT IN' -%}
        col2 NOT IN {{ filter.get('val')|where_in }}
    {% elif filter.get('op') == 'ILIKE' -%}
        LOWER(col2) LIKE LOWER({{ filter.get('val') }})
    {% elif filter.get('op') == 'IS NULL' -%}
        col2 IS NULL
    {% elif filter.get('op') == 'IS NOT NULL' -%}
        col2 IS NOT NULL
    {% else -%}
        col2 {{ filter.get('op') }} {{ filter.get('val') }}
    {%- endif -%}
{%- endfor %}

-- GROUP BY
  
{% set ns = namespace(groupByEmitted=false) %}

{% for col in finalColumns %}

	{% if col == 'col1' or col == '__ALL__' or (col is mapping and 'sqlExpression' in col and col.sqlExpression == 'col1') %}
		{%- if not ns.groupByEmitted %}
			GROUP BY
			{% set ns.groupByEmitted = true %}
		{% endif -%}
		col1
		{%- if col == '__ALL__' -%},{%- endif -%}
		{%- if not loop.last -%},{%- endif -%}
	{% endif %}
	
	{% if col == 'col2' or col == '__ALL__' or (col is mapping and 'sqlExpression' in col and col.sqlExpression == 'col2') %}
		{%- if not ns.groupByEmitted %}
			GROUP BY
			{% set ns.groupByEmitted = true %}
		{% endif -%}
		col2
		{%- if not loop.last -%},{%- endif -%}
	{% endif %}

{% endfor %}
  
``` 

3. Click "Create Chart"
4. Set a filter on one of the columns but don't select it as a dimension:
<img width="325" alt="image" src="https://github.com/user-attachments/assets/e9fe440b-6a31-4cd0-9a97-e1e1967a4e9b" />

5. On master, this breaks due to the following resulting SQL:

```
SELECT `measure` AS `measure`, `col2` AS `col2` 
FROM (
  SELECT
    sum(measure) as measure,
    col2
    
  FROM UNNEST([
    STRUCT(1 as measure, 'value1A' as col1, 'value2A' as col2),
    STRUCT(2 as measure, 'value1A' as col1, 'value2B' as col2),
    STRUCT(5 as measure, 'value1B' as col1, 'value2B' as col2),
    STRUCT(15 as measure, NULL as col1, NULL as col2)
  ])
	
  GROUP BY		
		col2
) AS `virtual_table` 
WHERE `col1` IS NULL
 LIMIT 1000
```

7. With this fix implemented, this works and the resulting SQL looks like this:

```
SELECT `measure` AS `measure`, `col2` AS `col2` 
FROM (

SELECT

  sum(measure) as measure,
  col2
  
FROM UNNEST([
  STRUCT(1 as measure, 'value1A' as col1, 'value2A' as col2),
  STRUCT(2 as measure, 'value1A' as col1, 'value2B' as col2),
  STRUCT(5 as measure, 'value1B' as col1, 'value2B' as col2),
  STRUCT(15 as measure, NULL as col1, NULL as col2)
])

 WHERE 1 = 1 
  AND
  col1 IS NULL
	
	GROUP BY	
		col2

) AS `virtual_table`
 LIMIT 1000
```



### ADDITIONAL INFORMATION
- [X] Fixes https://github.com/apache/superset/issues/33294
- [X] Required feature flags: ENABLE_TEMPLATE_PROCESSING = True

Prokos committed:
Fix get_filters for null operators

Prokos committed:
Fixed formatting of file through ruff-format

Prokos commented:
@sfirke Pre-commit now runs successfully locally :)

Prokos commented on superset/jinja_context.py:
```diff
@@ -378,7 +378,15 @@ def get_filters(self, column: str, remove_filter: bool = False) -> list[Filter]:
+                        FilterOperator.IS_NULL.value,
+                        FilterOperator.IS_NOT_NULL.value,
+                    )
+                )
```
Not me talking to an AI: after the ruff-format fix maybe the final logic is a bit too expanded but I'll leave it to a reviewer to decide on what formatting should be used. At least now it's passing the test.

sfirke commented:
I tagged a few committers who I believe are familiar with the Jinja internals. Hopefully one of them can take a quick look and we can get this merged!

Vitor-Avila commented on superset/jinja_context.py:
```diff
@@ -378,7 +378,16 @@ def get_filters(self, column: str, remove_filter: bool = False) -> list[Filter]:
                 and flt.get("clause") == "WHERE"
                 and flt.get("subject") == column
-                and val
+                and (
+                    val
+                    or
+                    # IS_NULL and IS_NOT_NULL operators do not have a value
+                    op
+                    in (
+                        FilterOperator.IS_NULL.value,
+                        FilterOperator.IS_NOT_NULL.value,
+                    )
+                )
```
hmm what do you think about this? I noticed the convo with Korbit too.
``` python
and (
    val
    # IS_NULL and IS_NOT_NULL operators do not have a value
    or op in (
        FilterOperator.IS_NULL.value,
        FilterOperator.IS_NOT_NULL.value,
    )
)
```

Prokos committed:
Changed formatting + added a test

Prokos commented:
@Vitor-Avila Thanks for having a look!

I added a test, that I confirmed fails without my change, and is successful with my change.

I've also implemented your formatting suggestion.

Vitor-Avila reviewed (approved):
LGTM! Thank you so much for fixing this flow!
